pkg/platform/model/auth.go

```diff
@@ -0,0 +1,60 @@
+package model
+
+import (
+	"time"
+
+	"github.com/go-openapi/strfmt"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/pkg/platform/api/mono"
+	"github.com/ActiveState/cli/pkg/platform/api/mono/mono_client/oauth"
+	"github.com/ActiveState/cli/pkg/platform/api/mono/mono_models"
+)
+
+// RequestDeviceAuthorization posts a request to authorize this device on the Platform and
+// returns the device code needed for authorization.
+// The user is subsequently required to visit the device code's URI and click the "Authorize"
+// button.
+func RequestDeviceAuthorization() (*mono_models.DeviceCode, error) {
+	postParams := oauth.NewAuthDevicePostParams()
+	response, err := mono.Get().Oauth.AuthDevicePost(postParams)
+	if err != nil {
+		logging.Error("Error requesting device authorization: %v", err)
+		return nil, locale.NewError("err_auth_device")
+	}
+	return response.Payload, nil
+}
+
+// WaitForAuthorization waits for the user to authorize a previously posted device authorization
+// request and returns the completed authorization.
+// Returns an error if authorization cannot be performed (e.g. timeout or Platform is unreachable).
+func WaitForAuthorization(deviceCodePayload *mono_models.DeviceCode) (*mono_models.DeviceCodeComplete, error) {
+	deviceCode := strfmt.UUID(*deviceCodePayload.DeviceCode)
+	getParams := oauth.NewAuthDeviceGetParams()
+	getParams.SetDeviceCode(deviceCode)
+	startTime := time.Now()
+	const timeout = 5 * 60 * time.Second
```

Comment:
Go has a cleaner pattern for this: https://gobyexample.com/timeouts

Reply:
I'm not sure if it will work for your use case but we have used a retry HTTP client before. It may require some special configuration and even then still might not work for what you want but could be worth a shot. It's called `retryhttp` and it's already in our vendor directory.

Reply:
I think retrying would only further complicate the timeout in this case though. And given the use-case retrying seems unnecessary.

Reply:
Pete suggested a timeout of 5 minutes in order to allow the user to complete the Platform sign up process if they have not already. The device authorization also has an internal timeout. The State Tool will poll for an affirmative authorization request until we either hit Pete's suggested timeout, or the Platform says we timed out, whichever comes first.

This is not an HTTP network timeout. Sorry if I'm stating the obvious.

Reply:
I don't think either points raised here have anything to do with my original comment, which is that Go has a pattern for handling timeouts that I feel we should adopt here. See the link in my original comment.

Reply:
I've just removed an explicit timeout from the client. The Platform has its own timeout and returns an "expired" status that is currently handled. There's no need to have another timeout mechanic.